eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala

```diff
@@ -137,28 +137,45 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: EclairWa
           stay() using d.copy(channels = d.channels + (TemporaryChannelId(temporaryChannelId) -> channel))
         }
 
       case Event(msg: protocol.OpenChannel, d: ConnectedData) =>
         d.channels.get(TemporaryChannelId(msg.temporaryChannelId)) match {
           case None =>
             val channelConfig = ChannelConfig.standard
-            val channelFeatures = ChannelFeatures.pickChannelFeatures(d.localFeatures, d.remoteFeatures)
-            val (channel, localParams) = createNewChannel(nodeParams, d.localFeatures, channelFeatures, funder = false, fundingAmount = msg.fundingSatoshis, origin_opt = None)
-            val temporaryChannelId = msg.temporaryChannelId
-            log.info(s"accepting a new channel with temporaryChannelId=$temporaryChannelId localParams=$localParams")
-            channel ! INPUT_INIT_FUNDEE(temporaryChannelId, localParams, d.peerConnection, d.remoteInit, channelConfig, channelFeatures)
-            channel ! msg
-            stay() using d.copy(channels = d.channels + (TemporaryChannelId(temporaryChannelId) -> channel))
+            val channelType_opt = msg.channelType_opt match {
+              case Some(proposedFeatures) =>
+                // We ensure that we support the features necessary for this channel type.
+                val featuresSupported = proposedFeatures.unknown.isEmpty && proposedFeatures.activated.forall { case (f, _) => d.localFeatures.hasFeature(f) }
+                if (featuresSupported) ChannelTypes.fromFeatures(proposedFeatures) else None
+              case None => Some(ChannelTypes.pickChannelType(d.localFeatures, d.remoteFeatures))
+            }
+            channelType_opt match {
+              case Some(channelType) =>
+                val (channel, localParams) = createNewChannel(nodeParams, d.localFeatures, channelType, funder = false, fundingAmount = msg.fundingSatoshis, origin_opt = None)
+                val temporaryChannelId = msg.temporaryChannelId
+                log.info(s"accepting a new channel with type=$channelType temporaryChannelId=$temporaryChannelId localParams=$localParams")
+                channel ! INPUT_INIT_FUNDEE(temporaryChannelId, localParams, d.peerConnection, d.remoteInit, channelConfig, channelType)
+                channel ! msg
+                stay() using d.copy(channels = d.channels + (TemporaryChannelId(temporaryChannelId) -> channel))
+              case None =>
+                msg.channelType_opt.foreach(proposedChannelType => {
+                  log.warning(s"ignoring open_channel with invalid channel type=$proposedChannelType")
+                  val ourDefaultChannelType = ChannelTypes.pickChannelType(d.localFeatures, d.localFeatures)
+                  d.peerConnection ! Error(msg.temporaryChannelId, InvalidChannelType(msg.temporaryChannelId, ourDefaultChannelType.features, proposedChannelType).getMessage)
+                })
+                stay()
+            }
```

Comment:
Here I would only deal with channel types, not features. Following up with my idea about `UnsupportedChannelTypes`, this is what I get:

```suggestion
            val channelType_either: Either[ChannelException, ChannelType] = msg.channelType_opt match {
              case Some(unsupportedChannelType: UnsupportedChannelType) =>
                  log.warning(s"ignoring open_channel with invalid channel type=$unsupportedChannelType")
                  val ourDefaultChannelType = ChannelTypes.pickChannelType(d.localFeatures, d.localFeatures)
                  Left(InvalidChannelType(msg.temporaryChannelId, ourDefaultChannelType.features, unsupportedChannelType).getMessage))
              case Some(channelType) => Right(channelType)
              case None => Right(ChannelTypes.pickChannelType(d.localFeatures, d.remoteFeatures))
            }
            channelType_either match {
              case Right(channelType) =>
                val (channel, localParams) = createNewChannel(nodeParams, d.localFeatures, channelType, funder = false, fundingAmount = msg.fundingSatoshis, origin_opt = None)
                val temporaryChannelId = msg.temporaryChannelId
                log.info(s"accepting a new channel with type=$channelType temporaryChannelId=$temporaryChannelId localParams=$localParams")
                channel ! INPUT_INIT_FUNDEE(temporaryChannelId, localParams, d.peerConnection, d.remoteInit, channelConfig, channelType)
                channel ! msg
                stay() using d.copy(channels = d.channels + (TemporaryChannelId(temporaryChannelId) -> channel))
              case Left(error) =>
                  d.peerConnection ! error
                stay()
            }
```

I'm not sure if we should include a channel type tlv in the `accept_channel`, if there was none in the `open_channel`. The spec does allow both behaviors.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1867/commits/03bf4c363efa65184c39a2af079fb8976f389b77
I think we should always send our `channel_type`, it's harmless and more explicit so it's always a win.